### PR TITLE
Refactor tool arguments usage to be implicit downcast safe

### PR DESCRIPTION
**Pull request by jonahwilliams**
## Description

Provides a typed wrapper for ArgResults that provides options (Strings), flags (bools), and multi-options (List\<String\>) as typed getters instead of using `operator []` to return dynamic.

## Related Issues

#13815

### CODE REVIEW

File: packages/flutter_tools/lib/src/base/args.dart [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+import 'package:args/args.dart';
+
+/// A typed wrapper for the flutter command [ArgResults].
+class TypedArgResults {
+  TypedArgResults(this._argResults);
+
+  final ArgResults _argResults;
+
+  /// Returns the value of the command line flag `name`.
+  bool getFlag(String name) => _argResults[name] as bool;
```
Comment (natebosch) [inline packages/flutter_tools/lib/src/base/args.dart]: drive by comment - consider naming these `flag`, `option`, and `multiOption`, or maybe with the prefix `read`?

https://www.dartlang.org/guides/language/effective-dart/design#avoid-starting-a-method-name-with-get

Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/base/args.dart]: Updated these to readFlag, readOption, readMultiOption. I can see just flag, option, and multiOption reading well - though I'm never sure about methods without verbs.

**Comment by jonahwilliams** [comment]
It turns out this change is hard to make, because I keep running into implicit downcast errors...

**Comment by jonahwilliams** [comment]
I made some copy paste errors in this code, will take a closer look and update when they're fixed

**Comment by jonahwilliams** [comment]
Identified the cause of the failing tests as 1) test cases that reuse the flutter command. This requires us to wrap each access of `argResults` with `typedArgs`, since the setter used is private. 2) A seemingly broken test case that was re-using an existing process manager. This caused it to accidentally pass, as the verify calls were correct wrt the last test case.

File: packages/flutter_tools/lib/src/base/args.dart [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+import 'package:args/args.dart';
+
+/// A typed wrapper for the flutter command [ArgResults].
+class TypedArgResults {
+  TypedArgResults(this._argResults);
+
+  final ArgResults _argResults;
+
+  /// Returns the value of the command line flag `name`.
+  bool readFlag(String name) => _argResults[name] as bool;
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/base/args.dart]: What do you think about:

```dart
Map<String, bool> _flags;
Map<String, bool> get flags {
  if (_flags == null) {
    _flags = Map.fromEntries(_argResults.entries.where((e) => e.value is bool);
  }
  return _flags;
}
```

File: packages/flutter_tools/lib/src/commands/analyze.dart [hunk 3]
```diff
@@ -72,7 +72,7 @@ class AnalyzeCommand extends FlutterCommand {
   @override
   bool get shouldRunPub {
     // If they're not analyzing the current project.
-    if (!argResults['current-package']) {
+    if (!args.readFlag('current-package')) {
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/commands/analyze.dart]: Then you've got:
```dart
args.flags['current-package']
```

File: packages/flutter_tools/lib/src/base/args.dart (hunk 2 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/base/args.dart]: My hope is that something like readFlag or flag is added by the args package (https://github.com/dart-lang/args/issues/95) and then we could drop the wrapper entirely.

**Comment by jonahwilliams** [comment]
I don't have time to get to this lately :(

**Pull request closed without merging**